src/plugins/vis_type_vega/public/line_vega_spec_expression.ts

```diff
@@ -0,0 +1,246 @@
+  visLayers: string | null;
+  visParams: string;
+  dimensions: string;
+}
+
+export type VegaSpecExpressionFunctionDefinition = ExpressionFunctionDefinition<
+  'vega_spec',
+  Input,
+  Arguments,
+  Output
+>;
+
+const createSpecFromDatatable = (
+  datatable: OpenSearchDashboardsDatatable,
+  visParams: string,
+  dimensionsString: string
+): object => {
+  // TODO: we can try to use VegaSpec type but it is currently very outdated, where many
+  // of the fields and sub-fields don't have other optional params that we want for customizing.
+  // For now, we make this more loosely-typed by just specifying it as a generic object.
+  const spec = {} as any;
+  const xAxis = datatable.columns[0];
+
+  const parseParams = JSON.parse(visParams);
+  const dimensions = JSON.parse(dimensionsString);
+  const legendPosition = parseParams.legendPosition;
+
+  // Get time range for the data in case there is only data for a small range so it will show the full time range
+  const startTime = {};
+  const xAxisId = xAxis.id.toString();
+  // @ts-ignore
+  startTime[xAxisId] = new Date(dimensions.x.params.bounds.min).valueOf();
+  const endTime = {};
+  // @ts-ignore
+  endTime[xAxisId] = new Date(dimensions.x.params.bounds.max).valueOf();
+  const updatedTable = datatable.rows.concat([startTime, endTime]);
+
+  // TODO: update this to v5 when available
+  spec.$schema = 'https://vega.github.io/schema/vega-lite/v4.json';
+  spec.data = {
+    values: updatedTable,
+  };
+  spec.config = {
+    view: {
+      stroke: null,
+    },
+    concat: {
+      spacing: 0,
+    },
+    // the circle timeline representing annotations
+    circle: {
+      color: 'blue',
+    },
+    // the vertical line when user hovers over an annotation circle
+    rule: {
+      color: 'red',
+    },
+    legend: {
+      orient: legendPosition,
+    },
+  };
+
+  // Get the valueAxes data and generate a map to easily fetch the different valueAxes data
+  const valueAxis = {};
+  parseParams.valueAxes.forEach((yAxis: { id: { toString: () => string | number } }) => {
+    // @ts-ignore
+    valueAxis[yAxis.id.toString()] = yAxis;
+  });
+
+  spec.layer = [] as any[];
+
+  if (datatable.rows.length > 0) {
+    let skip = 0;
+    datatable.columns.forEach((column, index) => {
+      const currentSeriesParams = parseParams.seriesParams[index - skip];
+      // Check if its not xAxis column data
+      if (column.meta?.aggConfigParams?.interval != null) {
+        skip++;
+      } else {
+        const currentValueAxis =
+          // @ts-ignore
+          valueAxis[currentSeriesParams.valueAxis.toString()];
+        let tooltip: Array<{ field: string; type: string; title: string }> = [];
+        if (parseParams.addTooltip) {
+          tooltip = [
+            { field: xAxis.id, type: 'temporal', title: xAxis.name.replaceAll('"', '') },
+            { field: column.id, type: 'quantitative', title: column.name.replaceAll('"', '') },
+          ];
+        }
+        spec.layer.push({
```

Comment:
I'm not sure we need separate layers for each column - I like the layer separation more for time markers and thresholds as you have them below. But maybe I don't quite have the right mental model for the datatable shapes. That's where unit tests may be helpful.

Reply:
I was having to do separate layers for each column due to them potentially being part of different axis and to get the legend to work correctly and allow many different y value metrics displayed on the same graph.